tests/test_datasets.py

```diff
@@ -0,0 +1,543 @@
+
+            assert load._reader._name == "pandas"
+            monkeypatch.delattr(altair.datasets, "load")
+
+            monkeypatch.setitem(sys.modules, "pandas", None)
+            monkeypatch.delitem(sys.modules, "pyarrow")
+            monkeypatch.setitem(sys.modules, "pyarrow", import_module("pyarrow"))
+            from altair.datasets import load
+
+            assert load._reader._name == "pyarrow"
+            monkeypatch.delattr(altair.datasets, "load")
+            monkeypatch.setitem(sys.modules, "pyarrow", None)
+
+            with pytest.raises(NotImplementedError, match="no.+backend"):
+                from altair.datasets import load
+
+
+@backends
+def test_loader_call(backend: _Backend, monkeypatch: pytest.MonkeyPatch) -> None:
+    monkeypatch.delenv(CACHE_ENV_VAR, raising=False)
+
+    data = Loader.with_backend(backend)
+    frame = data("stocks", ".csv")
+    assert is_into_dataframe(frame)
+    nw_frame = nw.from_native(frame)
+    assert set(nw_frame.columns) == {"symbol", "date", "price"}
+
+
+@backends
+def test_missing_dependency_single(
+    backend: _Backend, monkeypatch: pytest.MonkeyPatch
+) -> None:
+    if backend in {"polars[pyarrow]", "pandas[pyarrow]"}:
+        pytest.skip("Testing single dependency backends only")
+
+    monkeypatch.setitem(sys.modules, backend, None)
+
+    with pytest.raises(
+        ModuleNotFoundError,
+        match=re.compile(
+            rf"{backend}.+requires.+{backend}.+but.+{backend}.+not.+found.+pip install {backend}",
+            flags=re.DOTALL,
+        ),
+    ):
+        Loader.with_backend(backend)
+
+
+@pytest.mark.parametrize("backend", ["polars[pyarrow]", "pandas[pyarrow]"])
+@skip_requires_pyarrow
+def test_missing_dependency_multi(
+    backend: _Backend, monkeypatch: pytest.MonkeyPatch
+) -> None:
+    secondary = "pyarrow"
+    primary = backend.removesuffix(f"[{secondary}]")
+    monkeypatch.setitem(sys.modules, secondary, None)
+
+    with pytest.raises(
+        ModuleNotFoundError,
+        match=re.compile(
+            rf"{re.escape(backend)}.+requires.+'{primary}', '{secondary}'.+but.+{secondary}.+not.+found.+pip install {secondary}",
+            flags=re.DOTALL,
+        ),
+    ):
+        Loader.with_backend(backend)
+
+
+@backends
+def test_dataset_not_found(backend: _Backend) -> None:
+    """
+    Various queries that should **always raise** due to non-existent dataset.
+
+    ``Loader.url`` is used since it doesn't require a remote connection.
+    """
+    import polars as pl
+
+    data = Loader.with_backend(backend)
+    real_name: Literal["disasters"] = "disasters"
+    real_suffix: Literal[".csv"] = ".csv"
+    real_tag: Literal["v1.14.0"] = "v1.14.0"
+
+    invalid_name: Literal["fake name"] = "fake name"
+    invalid_suffix: Literal["fake suffix"] = "fake suffix"
+    invalid_tag: Literal["fake tag"] = "fake tag"
+
+    incorrect_suffix: Literal[".json"] = ".json"
+    incorrect_tag: Literal["v1.5.0"] = "v1.5.0"
```

Comment:
> A new separate point: I notice in the docstring for the url class that it requires a backend to be set (or at least it seems like that from the example). Would it be possible to make the URL accessible without having any dataframe package installed since Altair can be used directly with the URL string even if pandas/polars etc are not installed?

@joelostblom https://github.com/vega/altair/pull/3631#issuecomment-2480832711

> Will defining the dataset by URL goes through "Loader"?

@mattijn  https://github.com/vega/altair/discussions/3150#discussioncomment-11280516

The tests within this function give some examples of how [`metadata.parquet`](https://github.com/vega/altair/pull/3631#discussion_r1846456757) is utilized to return a url string.

Some key parts to the implementation:

<details><summary><code>_readers._Reader.url</code></summary>
<p>

<details><summary>Docs</summary>
<p>

https://github.com/vega/altair/blob/7ddb2a8c1e8ec6477cfc646c385e0b168f2fd330/altair/datasets/__init__.py#L243-L287

</p>
</details> 

https://github.com/vega/altair/blob/7ddb2a8c1e8ec6477cfc646c385e0b168f2fd330/altair/datasets/_readers.py#L160-L173

</p>
</details> 

<details><summary><code>_readers.validate_constraints</code></summary>
<p>

Strict equality filters are built up, in some cases, raising before even attempting to read [`metadata.parquet`](https://github.com/vega/altair/pull/3631#discussion_r1846456757)

https://github.com/vega/altair/blob/7ddb2a8c1e8ec6477cfc646c385e0b168f2fd330/altair/datasets/_readers.py#L441-L462

</p>
</details> 

If we were to change this, I would want to preserve the following benefits it currently brings:

### Small file size (**~18KB**)/fast performance
- We *could* split the file into multiple `.(csv|json)`(s)
	- E.g. instead of a flat file with redundant data, more like a normalized database schema
- Would require performing multiple table joins
	- Still needs a dataframe backend
	- Alternative would be `json.load` and using `set` operations for similar results

### Always returning a url that actually exists
  - The method doesn't require a remote connection
  - Generally aiming to provide error messages that are more informative than `URLError`
  - The guarantee is provided during generation in

https://github.com/vega/altair/blob/7ddb2a8c1e8ec6477cfc646c385e0b168f2fd330/tools/datasets/github.py#L375
  

### Shares much of the same code as local/remote file reading
Currently, all of these work in very similar ways - which I think will be helpful in maintaining it going forward.
While there may be concepts might are unfamiliar to some; understanding how any one of these paths works help with others.

https://github.com/vega/altair/blob/7ddb2a8c1e8ec6477cfc646c385e0b168f2fd330/altair/datasets/_readers.py#L133

https://github.com/vega/altair/blob/7ddb2a8c1e8ec6477cfc646c385e0b168f2fd330/altair/datasets/_readers.py#L160

https://github.com/vega/altair/blob/7ddb2a8c1e8ec6477cfc646c385e0b168f2fd330/altair/datasets/_readers.py#L175

Reply:
### Note
Need to investigate `stdlib` options

- https://docs.python.org/3/library/archiving.html
- https://docs.python.org/3/library/shutil.html#module-shutil